### PR TITLE
Add security considerations section on data transformations.

### PR DESCRIPTION
**Pull request by msporny**
This PR adds a security considerations section on data transforms and requests feedback on collision-resistant canonicalization requirements and asks if documentation should be provided on how modern canonicalization schemes avoid the pitfalls of XML Canonicalization.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/pull/42.html" title="Last updated on Aug 27, 2022, 7:36 PM UTC (3fc9070)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/42/b7cee0f...3fc9070.html" title="Last updated on Aug 27, 2022, 7:36 PM UTC (3fc9070)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
...
+        </p>
+        <p class="issue"
+          title="Avoiding the pitfalls of XML Canonicalization">
+The VCWG is seeking feedback on whether to explain why modern canonicalization
+schemes are simpler than the far more complex XML Canonicalization schemes of
+the early 2000s. Some readers seem to be under the impression that all
+canonicalization is difficult and has to be avoided at all costs. The WG would
+like to understand if a section explaining how some simpler data syntaxes (such
+as JSON) are easier to canonicalize than more complex data syntaxes (such as
+XML).
```
Comment (dlongley) [inline index.html]:
```suggestion
canonicalization is difficult and has to be avoided at all costs (including costs
to application developers). The WG would like to understand if a section
explaining why some simpler data syntaxes (such as JSON) are easier to
canonicalize than more complex data syntaxes (such as XML).
```

File: index.html [hunk 2]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
         </p>
       </section>
 
+      <section>
+        <h3>Transformations</h3>
+
+        <p>
+At times, it becomes important to transform the data being protected during the
+cryptographic protection process. Transformations can enable a particular type
+of cryptographic protection to be agnostic to the data format it is carried in.
```
Comment (TallTed) [inline index.html]:
```suggestion
At times, it is beneficial to transform the data being protected during the
cryptographic protection process. Such "in-line" transformation can enable a particular type
of cryptographic protection to be agnostic to the data format it is carried in.
```

File: index.html [hunk 3]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
...
+For example, some Data Integrity cryptographic suites utilize RDF Dataset
+Canonicalization [[?URDNA2015]] which transforms the representation into a
```
Comment (TallTed) [inline index.html]:
```suggestion
Canonicalization [[?URDNA2015]] which transforms the initial representation into a
```

File: index.html [hunk 4]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
...
+the same digital signature over the information to be expressed in JSON, CBOR,
+YAML, and other compatible syntaxes without having to recreate the cryptographic
+proof for every syntax.
+        </p>
+        <p>
+Being able to express the same digital signature across a variety of syntaxes is
+beneficial because systems often have native data formats with which they
+operate. For example, some systems are written against JSON data, while others
+are written against CBOR data. Without transformation, systems that process
+their data as CBOR internally are required to store the digitally signed data
```
Comment (TallTed) [inline index.html]:
```suggestion
their data internally as CBOR are required to store the digitally signed data
```

File: index.html [hunk 5]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
...
+structures as JSON (or vice-versa). This leads to double-storing data and can
+lead to extra security attack surface if the unsigned representation stored in
+databases accidentally deviates from the signed representation. By using
+transformations, the digital proof can live with the native data format to
```
Comment (TallTed) [inline index.html]:
```suggestion
transformations, the digital proof can live in the native data format to
```

File: index.html [hunk 6]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
...
+prevent undetectable database drift over time.
```
Comment (TallTed) [inline index.html]:
```suggestion
help prevent otherwise undetectable database drift over time.
```

File: index.html [hunk 7]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
...
+        </p>
+        <p>
+This specification is designed to avoid requiring the duplication of signed
+information by utilizing data transformations. Application developers are urged
```
Comment (TallTed) [inline index.html]:
```suggestion
information by utilizing "in-line" data transformations. Application developers are urged
```

File: index.html [hunk 8]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
...
+transformation mechanisms that can map to the same output. That is, given
+the different inputs for canonicalization scheme #1 and canonicalization
+scheme #2, they must not produce the same output value. As an analogy, this is
+the same requirement for cryptographic hashing mechanisms and is why those
+schemes are designed to be collision resistant. Cryptographic canonicalization
+mechanisms have the same requirement. At present, this isn't a problem because
+the three expected canonicalization schemes, the Universal RDF Dataset
+Canonicalization Algorithm 2015 [[?URDNA2015]], JSON Canonicalization Scheme
+[[?RFC8785]], and a theoretical future base-encoding canonicalization have
+entirely different outputs.
```
Comment (TallTed) [inline index.html]:
```suggestion
the three expected canonicalization schemes &mdash; the Universal RDF Dataset
Canonicalization Algorithm 2015 [[?URDNA2015]], JSON Canonicalization Scheme
[[?RFC8785]], and a theoretical future base-encoding canonicalization &mdash; have
entirely different outputs.
```

File: index.html (hunk 1 above)
Comment (TallTed) [reply index.html]: `explaining how some simpler` in line 2070 (unedited) or line 2071 (@dlongley's edit) should be `explaining why some simpler`

**Review by TallTed — CHANGES_REQUESTED** [review]

**Review by dlongley — APPROVED** [review]

File: index.html [hunk 9]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
...
+At times, it is beneficial to transform the data being protected during the
+cryptographic protection process. Such "in-line" transformation can enable a particular type
+of cryptographic protection to be agnostic to the data format it is carried in.
+For example, some Data Integrity cryptographic suites utilize RDF Dataset
+Canonicalization [[?URDNA2015]] which transforms the initial representation into a
+canonical form [[?N-QUADS]] that is then serialized, hashed, and digitally
+signed. As long as any syntax expressing the protected data can be transformed
+into this canonical form, the digital signature can be verified. This enables
+the same digital signature over the information to be expressed in JSON, CBOR,
+YAML, and other compatible syntaxes without having to recreate the cryptographic
```
Comment (TallTed) [inline index.html]:
```suggestion
YAML, and other compatible syntaxes without having to create a cryptographic
```

File: index.html [hunk 10]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
...
+proof for every syntax.
+        </p>
+        <p>
+Being able to express the same digital signature across a variety of syntaxes is
+beneficial because systems often have native data formats with which they
+operate. For example, some systems are written against JSON data, while others
+are written against CBOR data. Without transformation, systems that process
+their data internally as CBOR are required to store the digitally signed data
+structures as JSON (or vice-versa). This leads to double-storing data and can
+lead to extra security attack surface if the unsigned representation stored in
```
Comment (TallTed) [inline index.html]:
```suggestion
lead to increased security attack surface if the unsigned representation stored in
```

File: index.html [hunk 11]
```diff
@@ -2003,6 +2009,70 @@ <h3>Agility vs. Hardening</h3>
...
+being protected. Developers are also urged to regularly confirm that the
+cryptographically protected data has not been tampered with as it is written to
+and read from application storage.
+        </p>
+        <p class="issue"
+          title="Collision-resistant canonicalization requirements">
+The VCWG is seeking feedback on normative language that cryptographic suite
+implementers need to follow to ensure that they do not utilize data
+transformation mechanisms that can map to the same output. That is, given
+the different inputs for canonicalization scheme #1 and canonicalization
```
Comment (TallTed) [inline index.html]:
```suggestion
different inputs for canonicalization scheme #1 and canonicalization
```

Comment (TallTed) [reply index.html]: Or do I misunderstand... Should pushing the *same* input through the two canonicalization schemes produce two different outputs?

**Review by TallTed — CHANGES_REQUESTED** [review]
Minor tweaks, and a possible misunderstanding

File: index.html (hunk 11 above)
Comment (msporny) [reply index.html]: Both are logically equivalent statements. I think you are just stating the logically polar opposite of what is in the spec right now? I think these are the two statements under evaluation:

1. Pushing different inputs through two canonicalization schemes MUST NOT produce the same output value. This is what is in the spec text, if this happens, you have successfully produced a collision.
2. Pushing the same input through two canonicalization schemes MUST produce different output values. This is what I think you're saying. If this happens, you have successfully produced a collision.

I've been staring at this now for 10 minutes and am not 100% sure that I haven't confused myself.

**Comment by msporny** [comment]
Editorial, multiple reviews, changes requested and made, no objections, merging.

**Pull request merged**